File: services/dotnetcore/Controllers/RootController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Threading;
using Microsoft.AspNetCore.Mvc;
using OpenTracing;
using OpenTracing.Tag;
using Flurl.Http;
using dotnetcore.Models;

namespace dotnetcore.Controllers
{
  [Route("")]
  public class RootController : Controller
  {
    private ITracer trace;

    public RootController(ITracer tracer)
    {
      this.trace = tracer;
    }
    // GET api/values
    [HttpGet]
    public async Task<Response> GetAsync()
    {
      Random r = new Random();
      var sleepTime = r.Next(0, 1000);
      var requestTime = r.Next(0, 4);

      using (IScope scope = trace.BuildSpan("sleep").StartActive(finishSpanOnDispose: true))
      {
        try
        {
          if (sleepTime > 900)
          {
            throw new ApplicationException("I cant wait that long!");
          }
          Thread.Sleep(sleepTime);
        }
        catch (Exception ex)
        {
          Tags.Error.Set(scope.Span, true);
          scope.Span.Log(ex.Message);
        }

        using (IScope scope2 = trace.BuildSpan("httpbinrequest").StartActive(finishSpanOnDispose: true))
        {
          await $"https://httpbin.org/delay/{requestTime}".GetAsync();
        }
      }

      using (IScope scope = trace.BuildSpan("sendresponse").StartActive(finishSpanOnDispose: true))
      {
        return new Response { sleep = sleepTime, requestTime = requestTime };
      }
    }
  }
}

File: services/java/src/main/java/springtracingdemo/RootController.java
package springtracingdemo;

import java.net.*;

import io.opentracing.Scope;
import io.opentracing.tag.Tags;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.concurrent.ThreadLocalRandom;

@RestController
public class RootController {

  @Autowired
  private io.opentracing.Tracer tracer;

  @RequestMapping("/")
  public Response get() throws Exception {

    int sleepTime = ThreadLocalRandom.current().nextInt(0, 1000);
    int requestTime = ThreadLocalRandom.current().nextInt(0, 4);

    try (Scope sleepSpan = tracer.buildSpan("sleep").startActive(true)) {
      try {
        if (sleepTime > 900) {
          throw new Exception("I cant wait that long!");
        }
        Thread.sleep(sleepTime);
      } catch (Exception e) {
        Tags.ERROR.set(sleepSpan.span(), true);
        sleepSpan.span().log(e.getMessage());
      }

      try (Scope requestSpan = tracer.buildSpan("httpbinrequest").startActive(true)) {
        URL url = new URL("http://httpbin.org/delay/" + requestTime);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("GET");
      }
    }
    
    try (Scope scope = tracer.buildSpan("sendresponse").startActive(true)) {
      return new Response(sleepTime, requestTime);
    }
  }
}
